File: tests/workers/test_invert_image.py
# Copyright 2021 Xilinx, Inc.
# Copyright 2022 Advanced Micro Devices, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import argparse
import base64
import json
import os

import cv2
import numpy as np
import pytest

import amdinfer
import amdinfer.testing

from helper import root_path, run_benchmark, run_benchmark_func


def compare_jpgs(test_image, reference_image):
    """
    Compare a base64 encoded JPG image string with a reference image (as a np
    buffer). Since JPGs are lossy, images that are "equal" may be slightly
    different. We use the Manhattan norm to estimate the relative difference
    between the images.

    Args:
        test_image (str): Base64-encoded string without the header
        reference_image (np.ndarray): The image to compare against
        shape (list, optional): Check that the labeled shape of the test image
            matches its actual decoded shape. Defaults to None.
    """
    decoded_image = base64.decodebytes(test_image)
    buffer = np.frombuffer(decoded_image, dtype=np.uint8)
    received_image = cv2.imdecode(buffer, cv2.IMREAD_UNCHANGED)
    diff = received_image.astype(np.int16) - reference_image.astype(np.int16)
    m_norm = np.sum(np.absolute(diff))  # Manhattan norm
    m_norm_per_pixel = m_norm / np.prod(reference_image.shape)
    assert m_norm_per_pixel < 2  # arbitrarily use a threshold of 2 for comparison


def construct_request(asTensor):
    image_path = amdinfer.testing.getPathToAsset("asset_dog-3619020_640.jpg")
    modelMetadata = None
    request = amdinfer.ImageInferenceRequest(image_path, modelMetadata, asTensor)

    image = cv2.imread(image_path)
    image = cv2.bitwise_not(image)

    return request, image


def validate(response, image):
    assert not response.isError(), response.getError()

    assert response.model == "invert_image"

    outputs = response.getOutputs()
    assert len(outputs) == 1

    output = outputs[0]
    if output.datatype == amdinfer.DataType.BYTES:
        data = output.getStringData()
        compare_jpgs(data, image)
        assert output.parameters.empty()
    else:
        assert output.shape == [*image.shape]
        assert output.datatype == amdinfer.DataType.UINT8
        data = output.getUint8Data()
        assert len(data) == image.size
        assert (data == image.flatten().tolist()).all()
        assert output.parameters.empty()

    return response


@pytest.mark.usefixtures("load")
class TestInvertImage:
    """
    Test the InvertImage worker
    """

    @staticmethod
    def get_config():
        model = "CPlusPlus"
        parameters = amdinfer.ParameterMap()
        parameters.put("model", "invert_image")
        return (model, parameters)

    def send_request(self, request, image):
        """
        Sends the given request to the server and asserts common checks

        Args:
            request (dict): request to send to the server
            image (np.ndarray): Golden image to check against
            check_asserts (bool): Verify image against golden

        Returns:
            Response: Response
        """

        try:
            response = self.rest_client.modelInfer(self.endpoint, request)
        except ConnectionError:
            pytest.fail(
                "Connection to the amdinfer server ended without response!", False
            )

        validate(response, image)

    def test_invert_image(self):
        """
        Send a request to InvertImage as tensor data
        """
        request, image = construct_request(True)
        self.send_request(request, image)

    @pytest.mark.benchmark(group="invert_image")
    def test_benchmark_invert_image(self, benchmark):
        options = {
            "model": self.model,
            "parameters": self.parameters,
            "type": "rest (pytest)",
            "config": "N/A",
        }
        run_benchmark_func(benchmark, self.test_invert_image, **options)

    @pytest.mark.benchmark(group="invert_image")
    def test_benchmark_invert_image_request(self, benchmark):
        request, _ = self.construct_request(True)
        options = {
            "model": self.model,
            "parameters": self.parameters,
            "type": "rest (pytest)",
            "config": "N/A",
        }
        run_benchmark(
            benchmark, "invert_image_0", self.rest_client.modelInfer, request, **options
        )


@pytest.mark.usefixtures("load")
class TestInvertImage2:
    """
    Test the InvertImage worker
    """

    @staticmethod
    def get_config():
        model = ["CPlusPlus", "CPlusPlus", "CPlusPlus"]
        parameters_0 = amdinfer.ParameterMap()
        parameters_0.put("model", "base64_decode")
        parameters_1 = amdinfer.ParameterMap()
        parameters_1.put("model", "invert_image")
        parameters_2 = amdinfer.ParameterMap()
        parameters_2.put("model", "base64_encode")
        return (model, [parameters_0, parameters_1, parameters_2])

    def send_request(self, request, image):
        """
        Sends the given request to the server and asserts common checks

        Args:
            request (dict): request to send to the server
            image (np.ndarray): Golden image to check against
            check_asserts (bool): Verify image against golden

        Returns:
            Response: Response
        """

        try:
            response = self.rest_client.modelInfer(self.endpoint, request)
        except ConnectionError:
            pytest.fail(
                "Connection to the amdinfer server ended without response!", False
            )

        validate(response, image)

    def test_invert_image(self):
        """
        Send a request to InvertImage as a base64-encoded image
        """

        request, image = construct_request(False)
        self.send_request(request, image)

    @pytest.mark.benchmark(group="invert_image")
    def test_benchmark_invert_image(self, benchmark):
        options = {
            "model": self.model,
            "parameters": self.parameters,
            "type": "rest (pytest)",
            "config": "N/A",
        }
        run_benchmark_func(benchmark, self.test_invert_image, **options)

    @pytest.mark.benchmark(group="invert_image")
    def test_benchmark_invert_image_request(self, benchmark):
        request, _ = self.construct_request(False)
        options = {
            "model": self.model,
            "parameters": self.parameters,
            "type": "rest (pytest)",
            "config": "N/A",
        }
        run_benchmark(
            benchmark, "invert_image_1", self.rest_client.modelInfer, request, **options
        )

File: tests/models/test_invert_image.cpp
// Copyright 2023 Advanced Micro Devices, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <memory>                 // for allocator, unique_ptr
#include <opencv2/core.hpp>       // for Mat, CV_32FC3
#include <opencv2/imgcodecs.hpp>  // for imread
#include <opencv2/imgproc.hpp>    // for cvtColor, COLOR_BGR2GRAY
#include <string>                 // for string, operator+
#include <thread>                 // for yield
#include <vector>                 // for vector

#include "amdinfer/amdinfer.hpp"                   // for InferenceRequestInput
#include "amdinfer/testing/get_path_to_asset.hpp"  // for getPathToAsset
#include "amdinfer/testing/gtest_fixtures.hpp"  // for AssertionResult, Message
#include "amdinfer/util/base64.hpp"
#include "amdinfer/util/containers.hpp"
#include "amdinfer/util/filesystem.hpp"  // for readFile

namespace amdinfer {

void test(const amdinfer::Client* client, const std::string& version) {
  const auto path = getPathToAsset("asset_dog-3619020_640.jpg");

  const std::string model = "invert_image";
  const std::vector<std::string> endpoints{"invert_image", "execute",
                                           "invert_image_postprocess"};

  EXPECT_TRUE(client->modelList().empty());

  client->modelLoad(model, {}, version);
  for (const auto& endpoint : endpoints) {
    EXPECT_TRUE(client->modelReady(endpoint, version));
  }

  auto img = cv::imread(path);
  auto golden_img = img;
  cv::bitwise_not(img, golden_img);
  auto golden_shape = golden_img.rows * golden_img.cols * 3;

  // read the image as a jpeg
  auto raw_data = util::readFile(path);
  auto encoded_data = util::base64Encode(std::move(raw_data));

  InferenceRequestInput input_0;
  input_0.setName("input0");
  input_0.setDatatype(DataType::Bytes);
  input_0.setShape({encoded_data.size()});
  input_0.setData(encoded_data.data());

  InferenceRequest request;
  request.addInputTensor(input_0);

  auto response = client->modelInfer(endpoints[0], request, version);

  const auto& outputs = response.getOutputs();
  EXPECT_EQ(outputs.size(), 1);
  const auto& output = outputs[0];
  const auto& output_shape = output.getShape();
  EXPECT_TRUE(output_shape.size() == 1);
  EXPECT_EQ(output.getDatatype(), DataType::Bytes);
  const auto* data = static_cast<char*>(output.getData());

  auto decoded_data = util::base64Decode(data, output_shape.at(0));
  std::vector<char> buffer(decoded_data.begin(), decoded_data.end());
  auto received_image = cv::imdecode(buffer, cv::IMREAD_UNCHANGED);
  ASSERT_FALSE(received_image.empty());

  cv::Mat diff;
  cv::absdiff(received_image, golden_img, diff);
  auto manhattan_norm = cv::sum(cv::sum(diff))[0];
  auto manhattan_norm_per_pixel = manhattan_norm / golden_shape;
  EXPECT_TRUE(manhattan_norm_per_pixel < 2);

  unloadModels(client, endpoints, version);

  while (!client->modelList().empty()) {
    std::this_thread::sleep_for(std::chrono::seconds(1));
  }
}

#ifdef AMDINFER_ENABLE_GRPC
// @pytest.mark.extensions(["tfzendnn"])
// NOLINTNEXTLINE(cert-err58-cpp, cppcoreguidelines-owning-memory)
TEST_F(GrpcFixture, InvertImage) { test(client_.get(), ""); }

// @pytest.mark.extensions(["tfzendnn"])
// NOLINTNEXTLINE(cert-err58-cpp, cppcoreguidelines-owning-memory)
TEST_F(GrpcFixture, InvertImageVersioned) { test(client_.get(), "1"); }
#endif

// @pytest.mark.extensions(["tfzendnn"])
// NOLINTNEXTLINE(cert-err58-cpp, cppcoreguidelines-owning-memory)
TEST_F(BaseFixture, InvertImage) {
  NativeClient client(&server_);
  test(&client, "");
}

// @pytest.mark.extensions(["tfzendnn"])
// NOLINTNEXTLINE(cert-err58-cpp, cppcoreguidelines-owning-memory)
TEST_F(BaseFixture, InvertImageVersioned) {
  NativeClient client(&server_);
  test(&client, "1");
}

#ifdef AMDINFER_ENABLE_HTTP
// @pytest.mark.extensions(["tfzendnn"])
// NOLINTNEXTLINE(cert-err58-cpp, cppcoreguidelines-owning-memory)
TEST_F(HttpFixture, InvertImage) { test(client_.get(), ""); }

// @pytest.mark.extensions(["tfzendnn"])
// NOLINTNEXTLINE(cert-err58-cpp, cppcoreguidelines-owning-memory)
TEST_F(HttpFixture, InvertImageVersioned) { test(client_.get(), "1"); }
#endif

}  // namespace amdinfer
